File: easy/1160/main.go
package main

import "fmt"

func main() {
	fmt.Println(countCharacters([]string{"cat", "bt", "hat", "tree"}, "atach"))
	fmt.Println(countCharacters([]string{"hello", "world", "leetcode"}, "welldonehoneyr"))
}

func countCharacters(words []string, chars string) int {
	// chars map

	chars_map := map[string]int{}

	for _, char := range chars {
		chars_map[string(char)]++
	}
	sum := 0

	// check every word
	for _, word := range words {
		// word map
		word_map := map[string]int{}
		for _, char := range word {
			word_map[string(char)]++
		}
		sum += len(word)

		for key, value := range word_map {
			if chars_map[key] < value {
				sum -= len(word)
				break
			}
		}
	}
	return sum

}

File: easy/1160/main.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
  int countCharacters(vector<string> &words, string chars) {
    // count word frequency
    int chars_freq[26] = {};

    for (auto c : chars) {
      chars_freq[c - 'a']++;
    }

    auto total = 0;

    for (auto word : words) {
      int current_word_freq[26] = {};

      auto can_flag = true;

      for (auto w : word) {
        current_word_freq[w - 'a']++;
      }

      for (auto i = 0; i < 26; i++) {
        if (current_word_freq[i] > chars_freq[i]) {
          can_flag = false;
          break;
        }
      }
      if (can_flag == true) {
        total += word.size();
      }
    }
    return total;
  }
};
